File: plugins/web/src/sitemap.rs
//! sitemap.xml / robots.txt helpers (W7).

use serde_json::{json, Value};

fn cell_str(v: &Value) -> String {
    match v {
        Value::Null => String::new(),
        Value::String(s) => s.clone(),
        Value::Bool(b) => b.to_string(),
        Value::Number(n) => n.to_string(),
        other => other.to_string(),
    }
}

fn xml_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&apos;")
}

/// Build a sitemap.xml document from rows.
///
/// Each row may use keys: `loc`/`url`/`路径`, optional `lastmod`/`更新`,
/// `changefreq`/`频率`, `priority`/`优先级`.
pub fn build_sitemap(base: &str, items: &Value) -> String {
    let base = base.trim_end_matches('/');
    let rows: Vec<Value> = match items {
        Value::Array(a) => a
            .iter()
            .map(|row| match row {
                Value::String(s) => json!({ "loc": s }),
                Value::Object(_) => row.clone(),
                other => json!({ "loc": cell_str(other) }),
            })
            .collect(),
        Value::Object(m) => {
            // Columnar GFM: { loc: [...], ... }
            let locs = m
                .get("loc")
                .or_else(|| m.get("路径"))
                .or_else(|| m.get("path"))
                .or_else(|| m.get("url"));
            match locs {
                Some(Value::Array(a)) => a
                    .iter()
                    .enumerate()
                    .map(|(i, loc)| {
                        let mut obj = serde_json::Map::new();
                        obj.insert("loc".into(), loc.clone());
                        for (k, v) in m {
                            if k == "loc" || k == "路径" || k == "path" || k == "url" {
                                continue;
                            }
                            if let Some(Value::Array(col)) = Some(v) {
                                if let Some(cell) = col.get(i) {
                                    obj.insert(k.clone(), cell.clone());
                                }
                            }
                        }
                        Value::Object(obj)
                    })
                    .collect(),
                _ => Vec::new(),
            }
        }
        _ => Vec::new(),
    };
    let mut out = String::from(
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n\
         <urlset xmlns=\"http://www.sitemaps.org/schemas/sitemap/0.9\">\n",
    );
    for row in &rows {
        let Some(obj) = row.as_object() else {
            continue;
        };
        let loc_raw = obj
            .get("loc")
            .or_else(|| obj.get("url"))
            .or_else(|| obj.get("路径"))
            .or_else(|| obj.get("path"))
            .map(cell_str)
            .unwrap_or_default();
        if loc_raw.is_empty() {
            continue;
        }
        let loc = if loc_raw.starts_with("http://") || loc_raw.starts_with("https://") {
            loc_raw
        } else if loc_raw.starts_with('/') {
            format!("{base}{loc_raw}")
        } else {
            format!("{base}/{loc_raw}")
        };
        out.push_str("  <url>\n");
        out.push_str(&format!("    <loc>{}</loc>\n", xml_escape(&loc)));
        if let Some(lm) = obj
            .get("lastmod")
            .or_else(|| obj.get("更新"))
            .map(cell_str)
            .filter(|s| !s.is_empty())
        {
            out.push_str(&format!("    <lastmod>{}</lastmod>\n", xml_escape(&lm)));
        }
        if let Some(cf) = obj
            .get("changefreq")
            .or_else(|| obj.get("频率"))
            .map(cell_str)
            .filter(|s| !s.is_empty())
        {
            out.push_str(&format!(
                "    <changefreq>{}</changefreq>\n",
                xml_escape(&cf)
            ));
        }
        if let Some(pr) = obj
            .get("priority")
            .or_else(|| obj.get("优先级"))
            .map(cell_str)
            .filter(|s| !s.is_empty())
        {
            out.push_str(&format!("    <priority>{}</priority>\n", xml_escape(&pr)));
        }
        out.push_str("  </url>\n");
    }
    out.push_str("</urlset>\n");
    out
}

/// Default robots.txt body allowing all and pointing at a sitemap URL.
pub fn build_robots(sitemap_url: Option<&str>) -> String {
    let mut out = String::from("User-agent: *\nAllow: /\n");
    if let Some(u) = sitemap_url.filter(|s| !s.is_empty()) {
        out.push_str(&format!("Sitemap: {u}\n"));
    }
    out
}

/// ABI helper returning `{ body }`.
#[allow(dead_code)]
pub fn robots_json(sitemap_url: Option<&str>, body: Option<&str>) -> Value {
    let text = match body.filter(|s| !s.is_empty()) {
        Some(b) => b.to_string(),
        None => build_robots(sitemap_url),
    };
    json!({ "body": text })
}

File: plugins/web/internal/sitemap/sitemap.go
// Package sitemap builds sitemap.xml and robots.txt helpers (W7 / W-G7).
package sitemap

import (
	"fmt"
	"strconv"
	"strings"
)

func cellStr(v any) string {
	switch t := v.(type) {
	case nil:
		return ""
	case string:
		return t
	case bool:
		return strconv.FormatBool(t)
	case float64:
		if t == float64(int64(t)) {
			return strconv.FormatInt(int64(t), 10)
		}
		return strconv.FormatFloat(t, 'f', -1, 64)
	default:
		return fmt.Sprint(t)
	}
}

func xmlEscape(s string) string {
	r := strings.NewReplacer(
		"&", "&amp;",
		"<", "&lt;",
		">", "&gt;",
		`"`, "&quot;",
		"'", "&apos;",
	)
	return r.Replace(s)
}

func rowsFromItems(items any) []map[string]any {
	switch t := items.(type) {
	case []any:
		out := make([]map[string]any, 0, len(t))
		for _, row := range t {
			switch r := row.(type) {
			case map[string]any:
				out = append(out, r)
			case string:
				out = append(out, map[string]any{"loc": r})
			default:
				out = append(out, map[string]any{"loc": cellStr(r)})
			}
		}
		return out
	case map[string]any:
		locs, ok := firstArray(t, "loc", "路径", "path", "url")
		if !ok {
			return nil
		}
		out := make([]map[string]any, 0, len(locs))
		for i, loc := range locs {
			obj := map[string]any{"loc": loc}
			for k, v := range t {
				if k == "loc" || k == "路径" || k == "path" || k == "url" {
					continue
				}
				if col, ok := v.([]any); ok && i < len(col) {
					obj[k] = col[i]
				}
			}
			out = append(out, obj)
		}
		return out
	default:
		return nil
	}
}

func firstArray(m map[string]any, keys ...string) ([]any, bool) {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			if a, ok := v.([]any); ok {
				return a, true
			}
		}
	}
	return nil, false
}

// BuildSitemap builds a sitemap.xml document from rows.
func BuildSitemap(base string, items any) string {
	base = strings.TrimRight(base, "/")
	rows := rowsFromItems(items)
	out := `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
`
	for _, obj := range rows {
		locRaw := cellStr(first(obj, "loc", "url", "路径", "path"))
		if locRaw == "" {
			continue
		}
		loc := locRaw
		if strings.HasPrefix(locRaw, "http://") || strings.HasPrefix(locRaw, "https://") {
			loc = locRaw
		} else if strings.HasPrefix(locRaw, "/") {
			loc = base + locRaw
		} else {
			loc = base + "/" + locRaw
		}
		out += "  <url>\n"
		out += fmt.Sprintf("    <loc>%s</loc>\n", xmlEscape(loc))
		if lm := cellStr(first(obj, "lastmod", "更新")); lm != "" {
			out += fmt.Sprintf("    <lastmod>%s</lastmod>\n", xmlEscape(lm))
		}
		if cf := cellStr(first(obj, "changefreq", "频率")); cf != "" {
			out += fmt.Sprintf("    <changefreq>%s</changefreq>\n", xmlEscape(cf))
		}
		if pr := cellStr(first(obj, "priority", "优先级")); pr != "" {
			out += fmt.Sprintf("    <priority>%s</priority>\n", xmlEscape(pr))
		}
		out += "  </url>\n"
	}
	out += "</urlset>\n"
	return out
}

func first(m map[string]any, keys ...string) any {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			return v
		}
	}
	return nil
}

// BuildRobots returns a default robots.txt body with optional Sitemap line.
func BuildRobots(sitemapURL string) string {
	out := "User-agent: *\nAllow: /\n"
	if strings.TrimSpace(sitemapURL) != "" {
		out += "Sitemap: " + strings.TrimSpace(sitemapURL) + "\n"
	}
	return out
}

// SitemapJSON is the ABI helper returning {"xml":"…"}.
func SitemapJSON(base string, items any) map[string]any {
	return map[string]any{"xml": BuildSitemap(base, items)}
}
